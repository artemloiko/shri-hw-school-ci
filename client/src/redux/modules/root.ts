import { combineReducers } from 'redux';
import { combineEpics, ActionsObservable, StateObservable } from 'redux-observable';
import { buildsDetailsReducer } from '../../reducers/buildsDetails';
import { buildsReducer } from '../../reducers/builds';
import settingsReducer, { settingsEpic, SettingsActions, setSettingsEpic } from './settings';
import { EpicDependencies } from 'redux/createStore';
import { catchError } from 'rxjs/operators';

export const rootReducer = combineReducers({
  builds: buildsReducer,
  settings: settingsReducer,
  buildsDetails: buildsDetailsReducer,
});

export type RootState = ReturnType<typeof rootReducer>;
export type RootAction = SettingsActions;

export const rootEpic = (
  action$: ActionsObservable<RootAction>,
  state$: StateObservable<RootState>,
  dependencies: EpicDependencies,
) =>
  combineEpics(settingsEpic, setSettingsEpic)(action$, state$, dependencies).pipe(
    catchError((error, source) => {
      console.error('Epic error');
      console.dir(error);
      return source;
    }),
  );
