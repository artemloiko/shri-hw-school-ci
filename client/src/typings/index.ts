export * from '../../../server/typings/storage.interfaces';
