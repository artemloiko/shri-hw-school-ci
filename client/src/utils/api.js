import axios from 'axios';

const axiosAPI = axios.create({
  baseURL: '/api',
});

class API {
  constructor(axiosInstance) {
    this.axiosInstance = axiosInstance;
  }

  async getSettings() {
    const response = await this.axiosInstance.get('/settings');
    return response.data?.data;
  }

  /**
   * @param {object} settingsDTO School CI settings
   * @param {string} settingsDTO.repoName
   * @param {string} settingsDTO.buildCommand
   * @param {string} settingsDTO.mainBranch
   * @param {number} settingsDTO.period
   */
  async setSettings(settingsDTO) {
    const response = await this.axiosInstance.post('/settings', settingsDTO);
    return response.data?.data;
  }

  async getBuildsList(offset = 0, limit = 25) {
    const response = await this.axiosInstance.get(`/builds?offset=${offset}&limit=${limit}`);
    return response.data?.data;
  }

  async addBuild(commitHash) {
    const response = await this.axiosInstance.post(`/builds/${commitHash}`);
    return response.data?.data;
  }

  async getBuildDetails(buildId) {
    const response = await this.axiosInstance.get(`/builds/${buildId}`);
    return response.data?.data;
  }

  async getBuildLog(buildId) {
    const response = await this.axiosInstance.get(`/builds/${buildId}/logs`);
    return response.data;
  }
}

const ApiInstance = new API(axiosAPI);

export default ApiInstance;
