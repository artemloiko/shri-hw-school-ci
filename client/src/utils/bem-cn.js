import { withNaming } from '@bem-react/classname';

export const cn = (blockName, props) => {
  const cnBem = withNaming({ e: '__', m: '_', v: '_' });
  const { mods = {}, mix = [], className = '' } = props;
  const cnMods = cnBem(blockName)(mods);
  let cnMix = '';
  if (mix.length) {
    cnMix = mix.map((mixBlock) => cnBem(blockName, mixBlock)()).join(' ');
  }
  return `${className} ${cnMods} ${cnMix}`.trim();
};
