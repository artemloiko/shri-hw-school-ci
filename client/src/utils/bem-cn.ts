import { withNaming } from '@bem-react/classname';

export interface CNMods {
  [mod: string]: string | boolean | undefined;
}

export interface CNProps<T extends CNMods = {}> {
  className?: string;
  mix?: string[];
  mods?: T;
}

export function cn<T extends CNMods>(blockName: string, props: CNProps<T>): string {
  const cnBem = withNaming({ e: '__', m: '_', v: '_' });
  const { mods = {}, mix = [], className = '' } = props;
  const cnMods = cnBem(blockName)(mods);
  let cnMix = '';
  if (mix.length) {
    cnMix = mix.map((mixBlock) => cnBem(mixBlock, blockName)()).join(' ');
  }
  return `${className} ${cnMods} ${cnMix}`.trim();
}
