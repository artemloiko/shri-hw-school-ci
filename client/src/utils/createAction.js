/**
 * Create FSA (flux standart action);
 * @param {string} type
 * @param {*} payload
 * @param {boolean} error
 * @param {*} meta
 */
export function createAction(type, payload, error, meta) {
  const FSA = { type };
  if (payload) FSA.payload = payload;
  if (error) FSA.error = error;
  if (meta) FSA.meta = meta;
  return FSA;
}
