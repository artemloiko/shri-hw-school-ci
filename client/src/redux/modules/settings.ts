import { AxiosError } from 'axios';
import { ConfigurationModel, ConfigurationDTO } from 'typings';
import { EpicDependencies } from 'redux/createStore';
import { RootState } from './root';

import { Epic } from 'redux-observable';
import { isOfType } from 'typesafe-actions';
import { from, of } from 'rxjs';
import { mergeMap, map, catchError, filter, take, switchMap, tap } from 'rxjs/operators';

// Actions
export const GET_SETTINGS = 'school-ci/settings/GET_SETTINGS';
export const GET_SETTINGS_SUCCESS = 'school-ci/settings/GET_SETTINGS_SUCCESS';
export const GET_SETTINGS_FAIL = 'school-ci/settings/GET_SETTINGS_FAIL';
export const SET_SETTINGS = 'school-ci/settings/SET_SETTINGS';
export const SET_SETTINGS_SUCCESS = 'school-ci/settings/SET_SETTINGS_SUCCESS';
export const SET_SETTINGS_FAIL = 'school-ci/settings/SET_SETTINGS_FAIL';
export const RESET_SETTINGS_ERROR = 'school-ci/settings/RESET_SETTINGS_ERROR';

export interface SettingsState extends Partial<ConfigurationModel> {
  isFetching: boolean;
  error?: string;
}

export const initialState: SettingsState = {
  isFetching: true,
};

// Reducer
export default function settingsReducer(
  state = initialState,
  action: SettingsActions,
): SettingsState {
  switch (action.type) {
    case GET_SETTINGS_SUCCESS:
      return { ...state, ...action.payload, isFetching: false };
    case GET_SETTINGS_FAIL:
      return { ...state, error: action.payload, isFetching: false };
    case SET_SETTINGS_SUCCESS:
      return { ...state, ...action.payload };
    case SET_SETTINGS_FAIL:
      return { ...state, error: action.payload };
    case RESET_SETTINGS_ERROR:
      return { ...state, error: undefined };
    default:
      return state;
  }
}
// Action creators
export interface GetSettingsRequestAction {
  type: typeof GET_SETTINGS;
}
export const getSettingsRequest = (): GetSettingsRequestAction => ({
  type: GET_SETTINGS,
});

export interface GetSettingsSuccessAction {
  type: typeof GET_SETTINGS_SUCCESS;
  payload: ConfigurationModel;
}
export const getSettingsSuccess = (data: ConfigurationModel): GetSettingsSuccessAction => ({
  type: GET_SETTINGS_SUCCESS,
  payload: data,
});

export interface GetSettingsFailAction {
  type: typeof GET_SETTINGS_FAIL;
  payload: string;
  error: boolean;
}
export const getSettingsFail = (errorMessage: string): GetSettingsFailAction => ({
  type: GET_SETTINGS_FAIL,
  payload: errorMessage,
  error: true,
});

export interface SetSettingsAction {
  type: typeof SET_SETTINGS;
  payload: ConfigurationDTO;
  meta: {
    callback: (success: boolean) => void;
  };
}
export function setSettings(
  settingsDTO: ConfigurationDTO,
  callback: (success: boolean) => void,
): SetSettingsAction {
  return { type: SET_SETTINGS, payload: settingsDTO, meta: { callback } };
}

export interface SetSettingsSuccessAction {
  type: typeof SET_SETTINGS_SUCCESS;
  payload: ConfigurationDTO;
}
export function setSettingsSuccess(settingsDTO: ConfigurationDTO): SetSettingsSuccessAction {
  return { type: SET_SETTINGS_SUCCESS, payload: settingsDTO };
}

export interface SetSettingsFailAction {
  type: typeof SET_SETTINGS_FAIL;
  payload: string;
  error: boolean;
}
export function setSettingsFail(errorMessage: string): SetSettingsFailAction {
  return { type: SET_SETTINGS_FAIL, payload: errorMessage, error: true };
}

export interface ResetSettingsErrorAction {
  type: typeof RESET_SETTINGS_ERROR;
}
export function resetSettingsError(): ResetSettingsErrorAction {
  return { type: RESET_SETTINGS_ERROR };
}

export type SettingsActions =
  | GetSettingsRequestAction
  | GetSettingsSuccessAction
  | GetSettingsFailAction
  | SetSettingsAction
  | SetSettingsFailAction
  | SetSettingsSuccessAction
  | ResetSettingsErrorAction;

export const settingsEpic: Epic<SettingsActions, SettingsActions, RootState, EpicDependencies> = (
  action$,
  state$,
  { api },
) =>
  action$.pipe(
    filter(isOfType(GET_SETTINGS)),
    take(1),
    mergeMap(() =>
      from(api.getSettings()).pipe(
        map((data) => getSettingsSuccess(data)),
        catchError((error: AxiosError) => of(getSettingsFail(error.message))),
      ),
    ),
  );

export const setSettingsEpic: Epic<
  SettingsActions,
  SettingsActions,
  RootState,
  EpicDependencies
> = (action$, state$, { api }) =>
  action$.pipe(
    filter(isOfType(SET_SETTINGS)),
    switchMap((action) =>
      from(api.setSettings(action.payload)).pipe(
        map((data) => {
          return setSettingsSuccess(data);
        }),
        catchError((error: AxiosError) => {
          const apiError = error?.response?.data?.error;
          const errorMessage: string = apiError?.errorCode || apiError?.message || error.message;
          return of(setSettingsFail(errorMessage));
        }),
        tap((resultingAction) => {
          action.meta.callback(resultingAction.type === SET_SETTINGS_SUCCESS);
        }),
      ),
    ),
  );
