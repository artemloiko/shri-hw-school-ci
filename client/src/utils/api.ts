import axios, { AxiosInstance } from 'axios';
import { ConfigurationDTO, StorageResponce, ConfigurationModel, BuildModel } from 'typings';

const axiosAPI = axios.create({
  baseURL: '/api',
});

export class StorageAPI {
  constructor(private axiosInstance: AxiosInstance) {
    this.axiosInstance = axiosInstance;
  }

  async getSettings() {
    const response = await this.axiosInstance.get<StorageResponce<ConfigurationModel>>('/settings');
    return response.data?.data;
  }

  async setSettings(settingsDTO: ConfigurationDTO) {
    await this.axiosInstance.post('/settings', settingsDTO);
  }

  async getBuildsList(offset = 0, limit = 25) {
    const response = await this.axiosInstance.get<StorageResponce<BuildModel[]>>(
      `/builds?offset=${offset}&limit=${limit}`,
    );
    return response.data?.data;
  }

  async addBuild(commitHash: string) {
    const response = await this.axiosInstance.post<StorageResponce<BuildModel>>(
      `/builds/${commitHash}`,
    );
    return response.data?.data;
  }

  async getBuildDetails(buildId: string) {
    const response = await this.axiosInstance.get<StorageResponce<BuildModel>>(
      `/builds/${buildId}`,
    );
    return response.data?.data;
  }

  async getBuildLog(buildId: string) {
    const response = await this.axiosInstance.get<StorageResponce<string>>(
      `/builds/${buildId}/logs`,
    );
    return response.data.data;
  }
}

const ApiInstance = new StorageAPI(axiosAPI);

export interface HttpError {
  errorCode: string;
  message: string;
  status: number;
}

export default ApiInstance;
