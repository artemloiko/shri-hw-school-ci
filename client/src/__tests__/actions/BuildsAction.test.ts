import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { mocked } from 'ts-jest/utils';
import {
  fetchBuildsListIfNeeded,
  updateBuildsList,
  fetchMoreBuilds,
} from '../../actions/BuildsAction';
import apiMock from '../../utils/api';
import { RootState } from 'redux/modules/root';
import { BuildModel } from 'typings';

jest.mock('../../utils/api');

const middlewares = [thunk];
const mockStore = configureMockStore<RootState>(middlewares);
const initialState: RootState = {
  settings: {
    isFetching: false,
    isLoaded: false,
  },
  builds: {
    isFetching: false,
    isLoaded: false,
    isFetchingMore: false,
    isFullListLoaded: false,
    buildsList: [],
  },
  buildsDetails: {},
};

describe('builds actions', () => {
  test('fetching builds list success', async () => {
    const store = mockStore(initialState);
    const data: BuildModel[] = [
      {
        id: '7e6a4f9b-fb65-489d-ad96-566f177f3479',
        configurationId: '2906e339-bc49-4e86-9e44-eda86bf0a971',
        buildNumber: 3,
        commitMessage: 'Add option to work with async fn',
        commitHash: 'a63f2b3',
        branchName: 'master',
        authorName: 'Artem Loiko',
        status: 'Success',
        start: '2020-04-10T06:58:06.766',
        duration: 4549,
      },
      {
        id: 'ba77d958-3ae8-4087-84be-8b1af5451e00',
        configurationId: '2906e339-bc49-4e86-9e44-eda86bf0a971',
        buildNumber: 2,
        commitMessage: 'Add option to work with async fn',
        commitHash: 'a63f2b3',
        branchName: 'master',
        authorName: 'Artem Loiko',
        status: 'Success',
        start: '2020-04-09T15:51:08.548',
        duration: 518,
      },
    ];
    mocked(apiMock.getBuildsList).mockImplementationOnce(() => Promise.resolve(data));

    await store.dispatch<any>(fetchBuildsListIfNeeded());

    expect(store.getActions()).toMatchSnapshot();
  });

  test('fetching builds list fail', async () => {
    const store = mockStore(initialState);
    const error = { message: 'Internal server error' };
    mocked(apiMock.getBuildsList).mockImplementationOnce(() => Promise.reject(error));

    await store.dispatch<any>(fetchBuildsListIfNeeded());

    expect(store.getActions()).toMatchSnapshot();
  });

  test('fetching more builds success', async () => {
    const store = mockStore({
      ...initialState,
      ...{
        builds: {
          isLoaded: true,
          isFetching: false,
          isFetchingMore: false,
          isFullListLoaded: false,
          buildsList: new Array(25), // empty array to simulate some loaded data
        },
      },
    });
    const data: BuildModel[] = [
      {
        id: 'e59f7715-8da8-4e46-9312-bcba770d2c0f',
        configurationId: '2906e339-bc49-4e86-9e44-eda86bf0a971',
        buildNumber: 1,
        commitMessage: 'Add option to work with async fn',
        commitHash: 'a63f2b3',
        branchName: 'master',
        authorName: 'Artem Loiko',
        status: 'Success',
        start: '2020-04-09T15:45:58.471',
        duration: 4934,
      },
    ];
    mocked(apiMock.getBuildsList).mockImplementationOnce(() => Promise.resolve(data));

    await store.dispatch<any>(fetchMoreBuilds());

    expect(store.getActions()).toMatchSnapshot();
  });

  test('update builds list', () => {
    const store = mockStore(initialState);
    const data: BuildModel = {
      id: 'b6eef618-014a-45da-a3a6-3c05851f65bc',
      configurationId: '2906e339-bc49-4e86-9e44-eda86bf0a971',
      buildNumber: 4,
      commitMessage: 'Add option to work with async fn',
      commitHash: 'a63f2b3',
      branchName: 'master',
      authorName: 'Artem Loiko',
      status: 'Waiting',
    };

    store.dispatch<any>(updateBuildsList(data));

    expect(store.getActions()).toMatchSnapshot();
  });
});
