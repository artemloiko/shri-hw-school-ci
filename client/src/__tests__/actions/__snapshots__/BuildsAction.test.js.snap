// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`builds actions fetching builds list fail 1`] = `
Array [
  Object {
    "type": "GET_BUILDS_LIST_REQUEST",
  },
  Object {
    "error": true,
    "payload": "Internal server error",
    "type": "GET_BUILDS_LIST_FAIL",
  },
]
`;

exports[`builds actions fetching builds list success 1`] = `
Array [
  Object {
    "type": "GET_BUILDS_LIST_REQUEST",
  },
  Object {
    "payload": Array [
      Object {
        "authorName": "Artem Loiko",
        "branchName": "master",
        "buildNumber": 3,
        "commitHash": "a63f2b3",
        "commitMessage": "Add option to work with async fn",
        "configurationId": "2906e339-bc49-4e86-9e44-eda86bf0a971",
        "duration": 4549,
        "id": "7e6a4f9b-fb65-489d-ad96-566f177f3479",
        "start": "2020-04-10T06:58:06.766",
        "status": "Success",
      },
      Object {
        "authorName": "Artem Loiko",
        "branchName": "master",
        "buildNumber": 2,
        "commitHash": "a63f2b3",
        "commitMessage": "Add option to work with async fn",
        "configurationId": "2906e339-bc49-4e86-9e44-eda86bf0a971",
        "duration": 518,
        "id": "ba77d958-3ae8-4087-84be-8b1af5451e00",
        "start": "2020-04-09T15:51:08.548",
        "status": "Success",
      },
    ],
    "type": "GET_BUILDS_LIST_SUCCESS",
  },
]
`;

exports[`builds actions fetching more builds success 1`] = `
Array [
  Object {
    "type": "LOAD_MORE_BUILDS_REQUEST",
  },
  Object {
    "payload": Array [
      Object {
        "authorName": "Artem Loiko",
        "branchName": "master",
        "buildNumber": 1,
        "commitHash": "a63f2b3",
        "commitMessage": "Add option to work with async fn",
        "configurationId": "2906e339-bc49-4e86-9e44-eda86bf0a971",
        "duration": 4934,
        "id": "e59f7715-8da8-4e46-9312-bcba770d2c0f",
        "start": "2020-04-09T15:45:58.471",
        "status": "Success",
      },
    ],
    "type": "LOAD_MORE_BUILDS_SUCCESS",
  },
]
`;

exports[`builds actions update builds list 1`] = `
Array [
  Object {
    "payload": Object {
      "authorName": "Artem Loiko",
      "branchName": "master",
      "buildNumber": 4,
      "commitHash": "a63f2b3",
      "commitMessage": "Add option to work with async fn",
      "configurationId": "2906e339-bc49-4e86-9e44-eda86bf0a971",
      "id": "b6eef618-014a-45da-a3a6-3c05851f65bc",
      "status": "Waiting",
    },
    "type": "GET_BUILDS_LIST_UPDATE",
  },
]
`;
